[4.0] Backend: add a note-field to plugins

hgh-esn opened the pull request:
**Pull Request for Issue # .**

[4.0] Backend: a note-field for plugins #24920

**Summary of Changes**
Backend: Setting up a "note" field for plugins

**Testing Instructions**
1. Apply this pull request (PR) e.g. by using the Patchtester component.
2. Go to the database view and fix the 2 errors for Joomla CMS in order to apply the schema update SQL:
![screen shot 2019-05-31 at 10 24 23](https://issues.joomla.org/uploads/1/0714340948ed1dd78ebc03ae9df3567b.png)

3. Go into the plugin-manager listview. Select one of the listed plugins.
At the right-bottom  of the plugin-formview, you find the new input-field "Note".
Fill in any `<------ text ------>` into the note-field.
Save the plugin.
When going back to the plugin-listview, the "note-text" must be shown under the the plugins name as

plugin-name
(Note: `<------ text ------>`)

**Expected result**
You will see the "Note" in two views.
a) plugin-listview: Here it is only shown under the Plugins name.
b) plugin-formview: Here it is shown and you have the possibility to create/update the 
`<------ text ------>`. Note !  The chars are limited to 255 max. as defined in the db-configuration.

**Actual result**
- / -

**Documentation Changes Required**
may be ?

**Additional Infos**
(from me, the developer)
If you only fill "spaces" into the notefield, they are not shown in the formview, of course as the are invisible. In the Listview you will see then only this:

**plugin-name**
(Note: )

That's not nice, but the Joomla testers wanted it that way, because the other notes in (modules, fields, etc.) work that way too. 
From my point of view it would be, in the context of a consistent presentation, to consider, whether to continue in the act. way or - in the case of "only entered spaces" - suppresses the note display in the respective listview, because then there is no really meaningful content. 

The solution to do it is in :
File: `administrator/components/com_plugins/tmpl/plugins/default.php`

change: `<?php if (!empty($item->note)) : ?>` to `<?php if (!empty($item->note) && trim($item->note) != '') : ?>`

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

hgh-esn committed:
Add files via upload

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

hgh-esn committed:
Add files via upload

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

hgh-esn committed:
Add files via upload

[4.0] Backend: a note-field for plugins #24920

ReLater commented:
Hello @hgh-esn 

It's a bit hard to follow your changes when I look into the tabulator "Files changed" on GitHub because you exchange the whole files via upload.

Normally one creates a custom branch based upon the current 4.0.0-dev branch. Then changes the code inside the custom branch and commits the changes to the 4.0.0-dev branch here.

See an example here what's the result: https://github.com/joomla/joomla-cms/pull/24968/files
Just changed lines are highlighted.

On the other hand: You don't have to create new prs if you have to make changes. You can commit additional changes to the same pr.

Maybe somebody can help you here how to do that if you tell us which tools you use (I only use the GUI of GitHub. Thus I can't help.)

But again: Just a private note from my side. I don't have to decide that.

Bakual commented:
@high-esn Check your IDE settings. We use tabs to indent code. Your IDE likely uses spaces which results in changes in every line instead of only those you touched.

richard67 commented:
Beside this, postgresql schema updates are wrong, they are same as the mysql files and so have wrong syntax e.g. for qouting.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24967">issues.joomla.org/tracker/joomla-cms/24967</a>.</sub>

hgh-esn committed:
Add files via upload

richard67 commented on administrator/components/com_plugins/tmpl/plugin/edit.php:
```diff
@@ -136,6 +136,14 @@
 									<?php echo $this->form->getInput('element'); ?>
 								</div>
 							</div>
+<?php /*HGH*/ ?>			<div class="control-group">
```
What shall this be good for?

richard67 replied:
I mean the `<?php /*HGH*/ ?>`

hgh-esn replied:
That was a comment to identify that I've changed it. In the latest file this is removed.

hgh-esn committed:
Add files via upload

hgh-esn committed:
Add files via upload

hgh-esn committed:
Add files via upload

hgh-esn committed:
Add files via upload

hgh-esn committed:
Add files via upload

hgh-esn committed:
Add files via upload

richard67 commented on installation/sql/postgresql/joomla.sql:
```diff
@@ -519,6 +519,7 @@ CREATE TABLE IF NOT EXISTS "#__extensions" (
   "checked_out_time" timestamp without time zone DEFAULT '1970-01-01 00:00:00' NOT NULL,
   "ordering" bigint DEFAULT 0,
   "state" bigint DEFAULT 0,
+  `note` varchar(255) NOT NULL,
```
Isn't it obvious by comparing with the line before that column names are quoted with double quotes in postgresql?

richard67 replied:
In addition: The "NOT NULL" is definitely wrong, it must be allowed to have no note.

hgh-esn replied:
I changed to double quotes (sorry for that) and the "NOT NULL" in the latest file-version.

richard67 commented on administrator/components/com_admin/sql/updates/postgresql/4.0.0-2019-05-20.sql:
```diff
@@ -0,0 +1 @@
+ALTER TABLE `#__extensions` ADD COLUMN `note` varchar(255) AFTER `state`;
```
Please look up in other schema updates for postrgesql the correct syntax. This here is mysql syntax. In postgresql, table and column names are quoted with double quotes. In aadditon, postgresql does not support `AFTER ...` when adding columns.

richard67 replied:
In addition, the statement in the schema update is not consistent with the statement in joomla.sql. In Joomla.sql is a "NOT NULL" and here in the schema update there is no "NOT NULL".

richard67 replied:
The "NOT NULL" in joomla.sql is definitely wrong, it must be allowed to have no note.

richard67 replied:
Still the wrong quotes for table and column names. It needs double quotes for postgresql.

richard67 replied:
And also still wrong "AFTER ..." at the end.

hgh-esn replied:
I'm working on. Don't hurry.

Quy commented on administrator/language/en-GB/en-GB.com_plugins.ini:
```diff
@@ -7,6 +7,8 @@ COM_PLUGINS="Plugins"
 COM_PLUGINS_ELEMENT_HEADING="Element"
 COM_PLUGINS_ERROR_FILE_NOT_FOUND="The file %s could not be found."
+COM_PLUGINS_FIELD_NOTE_LABEL="Note <span class=small>(255 chars max. )</span>"
+COM_PLUGINS_FIELD_NOTE_DESC="an optional 'note' for this plugin."																 
```
```suggestion													 
```

Quy commented on administrator/components/com_plugins/tmpl/plugins/default.php:
```diff
@@ -111,6 +111,14 @@
 								<?php $editIcon = $item->checked_out ? '' : '<span class="fa fa-pen-square mr-2" aria-hidden="true"></span>'; ?>
 								<a class="hasTooltip" href="<?php echo Route::_('index.php?option=com_plugins&task=plugin.edit&extension_id=' . (int) $item->extension_id); ?>" title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo $this->escape(addslashes($item->name)); ?>">
 									<?php echo $editIcon; ?><?php echo $item->name; ?></a>
+									<?php	if ( $item->note ) : ?>
```
```suggestion
									<?php if ( $item->note ) : ?>
```

hgh-esn committed:
Add files via upload

richard67 commented on installation/sql/mysql/joomla.sql:
```diff
@@ -506,6 +506,7 @@ CREATE TABLE IF NOT EXISTS `#__extensions` (
   `checked_out_time` datetime NOT NULL DEFAULT '0000-00-00 00:00:00',
   `ordering` int(11) DEFAULT 0,
   `state` int(11) DEFAULT 0,
+  `note` varchar(255) NOT NULL,
```
The "NOT NULL" is definitely wrong, it must be allowed to have no note.

richard67 commented on layouts/joomla/edit/global.php:
```diff
@@ -27,20 +27,35 @@
-	'note',
-	'version_note',
-);
+		if ($component !== 'com_plugins')
```
Indentation of the complete if statement from here until the closing `}` in line 58 is one tab too much.

richard67 replied:
Or even 2 tabs too much.

hgh-esn committed:
Add files via upload

hgh-esn committed:
Add files via upload

hgh-esn committed:
Add files via upload

richard67 commented:
Why don't you just fix things and then when ready upload it at once with 1 commit for all? Why for every little change 1 comming with same message "Add files via upload"?
I hope that if this PR ever gets merged, then it will be merged with 1 squash commit so we don't see the 9845 (or so) commits "Add files via upload" which it will have at the end in Joomla history for the next 5 years.

hgh-esn committed:
Add files via upload

hgh-esn committed:
Add files via upload

hgh-esn committed:
Dropping COM_PLUGINS_FIELD_NOTE_DESC

hgh-esn committed:
changing <b>...></b>  to  >strong> .... </strong> and a space before concat operator

SniperSister commented on administrator/components/com_plugins/tmpl/plugins/default.php:
```diff
@@ -111,6 +111,14 @@
+									<div class="small">
+										<a class="hasTooltip" href="#"
+											title="<?php echo JText::_('JFIELD_NOTE_LABEL'); ?>">
+											<?php echo '<strong>Note:</strong> ' . $item->note; ?>
```
The hardcoded „Note:“ label has to be moved into a language string, the note itself needs escaping.

SniperSister replied:
The note is still not escaped, use $this->escape()

hgh-esn replied:
Sorry, but I changed it   see last commit: https://github.com/joomla/joomla-cms/pull/24967/commits/e64772fa211ed77bb4fdb09e572d10956a1f35e5

Quy replied:
I don't see the change `$this->escape()`.

hgh-esn replied:
I changed the hardcoded  '<strong>Note:</strong> '  string to html  and a language-key as:
`<strong><?php echo JText::_('JFIELD_NOTE_LABEL')  .  ': '; ?></strong> ....`
This better separates the html-statement from the text. Thus, a translation of the text "Note" into the national language is possible.

`$this->escape()`  ?? 
I didn't see the reason why -  can you please explain more?

SniperSister replied:
The note is user-provided content and needs to be escaped to prevent Cross-Site-Scripting security issues.

hgh-esn replied:
Ah, that's it - thanks.
I read something about that in your [book](https://books.google.de/books?id=zbBPAgAAQBAJ&pg=PA221&lpg=PA221&dq=$this-+escape+joomla&source=bl&ots=35Z42LtJoi&sig=ACfU3U1nWVJTYZp61s4dREGJl98tKJ9hiw&hl=de&sa=X&ved=2ahUKEwjntrvZwLLiAhXMAewKHbE_BYcQ6AEwC3oECDAQAQ#v=onepage&q=%24this-%20escape%20joomla&f=false).

Is that the right way?
`<strong><?php echo JText::_('JFIELD_NOTE_LABEL')  .  ': ';  ?></strong><?php  echo $this->escape($item->note); ?>`

SniperSister replied:
Yes, that’s the right way :)

hgh-esn replied:
Ok.
I never have thought about it in anyway. Does that mean, that I have to change up all my priv. plugins, where users can enter data in fields. Just to prevent for Cross-Site-Scripting?
I think this would be a good way to do. 

I'll do the change here tomorrow. By for now.

hgh-esn replied:
I've changed it now.

hgh-esn committed:
separating html tag: <strong> ... </strong> from language string

hgh-esn committed:
Note-string is now escaped, delete language note-string ...

Delete added note-string in  ../administrator/language/en-GB.com_plugins.ini
Now using existing JFIELD_NOTE_LABEL in ../administrator/language/en-GB/en-GB.ini

hgh-esn committed:
[4.0] Note-string is now escaped

hgh-esn committed:
[4.0] Note-string is now escaped

hgh-esn commented:
why this:  --- continuous-integration/drone/pr — Build is failing

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

Quy commented on administrator/components/com_plugins/tmpl/plugins/default.php:
```diff
@@ -111,6 +111,14 @@
 								<?php $editIcon = $item->checked_out ? '' : '<span class="fa fa-pen-square mr-2" aria-hidden="true"></span>'; ?>
 								<a class="hasTooltip" href="<?php echo Route::_('index.php?option=com_plugins&task=plugin.edit&extension_id=' . (int) $item->extension_id); ?>" title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo $this->escape(addslashes($item->name)); ?>">
 									<?php echo $editIcon; ?><?php echo $item->name; ?></a>
+									<?php if ( $item->note ) : ?>
```
```suggestion
								<?php if ( $item->note ) : ?>
```

Quy commented on administrator/components/com_plugins/tmpl/plugins/default.php:
```diff
@@ -111,6 +111,14 @@
+									<div class="small">
+										<a class="hasTooltip" href="#"
+											title="<?php echo JText::_('JFIELD_NOTE_LABEL'); ?>">
+											<strong><?php echo JText::_('JFIELD_NOTE_LABEL') . ': ';?></strong><?php  echo $item->note; ?>
```
```suggestion
											<strong><?php echo JText::_('JFIELD_NOTE_LABEL') . ': '; ?></strong><?php  echo $item->note; ?>
```

Quy commented on administrator/components/com_plugins/tmpl/plugins/default.php:
```diff
@@ -111,6 +111,14 @@
+									<?php if ( $item->note ) : ?>
+									<div class="small">
+										<a class="hasTooltip" href="#"
+											title="<?php echo JText::_('JFIELD_NOTE_LABEL'); ?>">
```
No need for the `title` attribute.

Quy commented on layouts/joomla/edit/global.php:
```diff
@@ -40,7 +42,20 @@
+else 
+{
+	$fields = $displayData->get('fields') ?: array(
+	'transition',
```
Indent lines 50-56.

hgh-esn replied:
Without this change/workaround, the note-field isn't displayed in the formview to enter data. Don't know why, but it is so.

Quy commented on layouts/joomla/edit/global.php:
```diff
@@ -27,7 +27,9 @@
+if ($component !== 'com_plugins')
+{
+	$fields = $displayData->get('fields') ?: array(
 	'transition',
```
Indent lines 33-44.

hgh-esn replied:
Without this change/workaround, the note-field isn't displayed in the formview to enter data. Don't know why, but it is so.

Quy replied:
The request is for code style to indent.

hgh-esn replied:
Did it for lines 33-44 and 50-56 ;)

Quy commented on administrator/components/com_plugins/tmpl/plugins/default.php:
```diff
@@ -111,6 +111,14 @@
+									<?php if ( $item->note ) : ?>
+									<div class="small">
+										<a class="hasTooltip" href="#"
+											title="<?php echo JText::_('JFIELD_NOTE_LABEL'); ?>">
```
Do you need tooltip here since it is the same as the anchor text?

hgh-esn replied:
Yes, I see the text "note:" for mobile-,tablet-computers and the additional hoovering tooltip for computers with mouse-support.
I also used "JFIELD_NOTE_LABEL" because of avoiding a new language-text. So I use the existing one. That avoids unnecessary  translations and avoids blowing up language-files. Looks later as something like this.
![image](https://user-images.githubusercontent.com/5299726/58261336-64ab3480-7d78-11e9-8d37-f9980aa1656c.png)

Quy replied:
But the tooltip adds no additional information so no point in having it.

hgh-esn replied:
I'll try to build it out.

hgh-esn replied:
Done

Quy commented on administrator/components/com_plugins/tmpl/plugins/default.php:
```diff
@@ -111,6 +111,14 @@
 								<?php $editIcon = $item->checked_out ? '' : '<span class="fa fa-pen-square mr-2" aria-hidden="true"></span>'; ?>
 								<a class="hasTooltip" href="<?php echo Route::_('index.php?option=com_plugins&task=plugin.edit&extension_id=' . (int) $item->extension_id); ?>" title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo $this->escape(addslashes($item->name)); ?>">
 									<?php echo $editIcon; ?><?php echo $item->name; ?></a>
+									<?php if ( $item->note ) : ?>
```
```suggestion
								<?php if ($item->note) : ?>
```

hgh-esn replied:
Don't know why it is here so. In the source-file it looks ok.

Quy commented on administrator/components/com_plugins/tmpl/plugins/default.php:
```diff
@@ -111,6 +111,14 @@
+									<div class="small">
+										<a class="hasTooltip" href="#"
+											title="<?php echo JText::_('JFIELD_NOTE_LABEL'); ?>">
+											<strong><?php echo JText::_('JFIELD_NOTE_LABEL') . ': ';?></strong><?php  echo $item->note; ?>
```
```suggestion
											<strong><?php echo Text::_('JFIELD_NOTE_LABEL') . ': '; ?></strong><?php  echo $item->note; ?>
```

hgh-esn replied:
The "<strong><?php echo **Text::**_('JFIELD_NOTE_LABEL')" isn't in the source. There I see only "JTEXT::"

hgh-esn committed:
$item->note escaped

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

hgh-esn committed:
indent lines:   33-44 and 50-56

hgh-esn committed:
removed tooltip  in lines 114 ... 118

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

hgh-esn commented:
Hello,

and what is going on next here?

Quy commented on administrator/components/com_plugins/tmpl/plugins/default.php:
```diff
@@ -111,6 +111,11 @@
 								<?php $editIcon = $item->checked_out ? '' : '<span class="fa fa-pen-square mr-2" aria-hidden="true"></span>'; ?>
 								<a class="hasTooltip" href="<?php echo Route::_('index.php?option=com_plugins&task=plugin.edit&extension_id=' . (int) $item->extension_id); ?>" title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo $this->escape(addslashes($item->name)); ?>">
 									<?php echo $editIcon; ?><?php echo $item->name; ?></a>
+									<?php if ( $item->note ) : ?>
```
```suggestion
									<?php if ($item->note) : ?>
```

hgh-esn replied:
I'll do it in this way:

`<?php if ($item->note)  :  ?>`    ->  `<?php if (!empty($item->note))  :  ?>`
 
The !empty(....) makes it more clearly to what is checked. 
Ok ?

hgh-esn replied:
Done.

franz-wohlkoenig commented:
@hgh-esn Release Lead @wilsonge decide if and when this get merged.

hgh-esn committed:
changed line 114 :  <?php if (!empty($item->note))  :  ?>

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

hgh-esn commented:
@franz-wohlkoenig 

... and who will do the test?

franz-wohlkoenig commented:
Everyone can do a test, as all are Volunteers.

hgh-esn commented:
@franz-wohlkoenig 
> Everyone can do a test, as all are Volunteers.

Please tell me more about the process.
I think the pr is coded and the checks are ok. How does it go from here?
a) Where is the test platform? Is this a separate environment or is the joomla4 alpha?
b) If so, how does the pr get in there?
c) Who decides (@wilsonge?), when the feature in joomla4 alpha? is taken over.
d) Do I have to do anything about this? Or is it enough to wait?

franz-wohlkoenig commented:
After 2 successfully Tests Pull Request get Status "Ready to Commit". Next Step Release Lead (@wilsonge) decide to merge or not.

hgh-esn committed:
update view to make it clearer (https://github.com/joomla/joomla-cms/issues/25021)

 The urls are now changed to text-links.
The first one opens a modal-window and the second one switches to an ext. website.

Quy commented:
Please separate the last commit to its own branch.

hgh-esn committed:
Delete file here it goes later in a new branch.

hgh-esn committed:
change: <?php if (!empty($item->note)) : ?>

So the note isn't displayed, when the user has entered only space-chars.

hgh-esn committed:
recreate file 

This file was deleted in the forked repro because of deleting a branch in joomla-cms

hgh-esn committed:
Added two messages for (#25021)

hgh-esn committed:
deleted of two changelog-messages

The must go into another branch .... for #25021

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

infograf768 commented on administrator/components/com_plugins/Model/PluginsModel.php:
```diff
@@ -213,7 +213,8 @@ protected function getListQuery()
 				'a.extension_id , a.name, a.element, a.folder, a.checked_out, a.checked_out_time,' .
-					' a.enabled, a.access, a.ordering'
+				' a.enabled, a.access, a.ordering,' .
+				' a.note'
```
add `' a.note'` on the same line as `' a.enabled, a.access, a.ordering'`, keeping the tab

hgh-esn replied:
@infograf768
I can't believe it.
There were no changes on it the last days. and until yesterday all checks were ok.
After merging with the latest Joomla4-branches this morning, this happens without anything doing by myself.
But I can do it and we see what happens.

richard67 replied:
@hgh-esn  No, these code style mistakes which @infograf768 suggests to change were in your PR since the beginning. I've suggested in a review how to change that. If you see that review on GitHub, there should be some button or option to accept the changes. Maybe that helps you.

infograf768 commented on administrator/language/en-GB/en-GB.com_plugins.ini:
```diff
@@ -7,6 +7,7 @@ COM_PLUGINS="Plugins"
 COM_PLUGINS_CONFIGURATION="Plugin: Options"
 COM_PLUGINS_ELEMENT_HEADING="Element"
 COM_PLUGINS_ERROR_FILE_NOT_FOUND="The file %s could not be found."
+COM_PLUGINS_FIELD_NOTE_LABEL="Note <span class=small>(255 chars max.)</span>"
```
No need to add `<span class=small>(255 chars max.)</span>`
We do not do it for other Notes.

hgh-esn replied:
@infograf768
I think it is a helpful info to the user, because the notefield  is no endless textfield. By definition in the database it's length is varchar(255).

richard67 replied:
@hgh-esn infograf768 is right. We do not do it for other notes fields. You should implement your new field in the same way as other note fields, everything else will end in inconsistend user interface.

richard67 commented on administrator/components/com_plugins/Model/PluginsModel.php:
```diff
@@ -213,7 +213,8 @@ protected function getListQuery()
 				'list.select',
 				'a.extension_id , a.name, a.element, a.folder, a.checked_out, a.checked_out_time,' .
-					' a.enabled, a.access, a.ordering'
+				' a.enabled, a.access, a.ordering,' .
```
```suggestion
					' a.enabled, a.access, a.ordering, a.note'
```

richard67 commented on administrator/components/com_plugins/Model/PluginsModel.php:
```diff
@@ -213,7 +213,8 @@ protected function getListQuery()
 				'a.extension_id , a.name, a.element, a.folder, a.checked_out, a.checked_out_time,' .
-					' a.enabled, a.access, a.ordering'
+				' a.enabled, a.access, a.ordering,' .
+				' a.note'
```
```suggestion

```

hgh-esn committed:
Apply suggestions from code review

corrected as suggested changes by @richard67

Co-Authored-By: Richard Fath <richard67@users.noreply.github.com>

richard67 commented on administrator/language/en-GB/en-GB.com_plugins.ini:
```diff
@@ -7,6 +7,7 @@ COM_PLUGINS="Plugins"
 COM_PLUGINS_CONFIGURATION="Plugin: Options"
 COM_PLUGINS_ELEMENT_HEADING="Element"
 COM_PLUGINS_ERROR_FILE_NOT_FOUND="The file %s could not be found."
+COM_PLUGINS_FIELD_NOTE_LABEL="Note <span class=small>(255 chars max.)</span>"
```
```suggestion
COM_PLUGINS_FIELD_NOTE_LABEL="Note"
```

hgh-esn committed:
removed the span-tag from COM_PLUGINS_FIELD_NOTE_LABEL

richard67 commented:
@franz-wohlkoenig There is no test item for this in the issue tracker. Any idea how to fix that? PR looks good to me now and I have just tested with success.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24967">issues.joomla.org/tracker/joomla-cms/24967</a>.</sub>

franz-wohlkoenig commented:
> There is no test item for this in the issue tracker. Any idea how to fix that?

Happens sometimes, don't know why but calling @mbabker helps always :-)

hgh-esn commented:
@richard67
Thanks to you. That was - in all - a strongly birth for this little change.
I felt as falling in every trap that was possible. ;)

But lately it is good to read that you tested it now successfully.

Quy commented on administrator/language/en-GB/en-GB.com_plugins.ini:
```diff
@@ -7,6 +7,7 @@ COM_PLUGINS="Plugins"
 COM_PLUGINS_CONFIGURATION="Plugin: Options"
 COM_PLUGINS_ELEMENT_HEADING="Element"
 COM_PLUGINS_ERROR_FILE_NOT_FOUND="The file %s could not be found."
+COM_PLUGINS_FIELD_NOTE_LABEL="Note"
```
Sort alpha order.

richard67 replied:
@hgh-esn For explanation: We always sort the language strings in 1 language file in alphabetical order. So Quy is right, you should move your string after line `COM_PLUGINS_FIELD_NAME_LABEL="Plugin Name"`

Quy commented on layouts/joomla/edit/global.php:
```diff
@@ -27,20 +27,35 @@
-);
+if ($component !== 'com_plugins')
+{
+	$fields = $displayData->get('fields') ?: array(
```
Use short array syntax. `array()` to `[]`

hgh-esn replied:
I think that is cosmetics. 
I only repeated that for my change and surrounding it in the if-switch, what I found in the file.

When to change, please send me an example, because I don't really know what you mean especially doing it with nested arrays.

hgh-esn replied:
Ok I read about it [here](https://www.php.net/manual/en/language.types.array.php)

When the only difference is to change the Array( .... )    to   Array[ ... ]   It looks easy, but I suggest there are more other traps.

But I think you will find a lot of examples for the "old" syntax in the Joomla files. So I think it would be a good solution to to change this global by an automatic process .

hgh-esn replied:
OK I figured it out
But it looks to me very strange, because it is much more unreadable as in the old syntax, especially in editors with syntax-highlighting.
new syntax:
`			 $fields = $displayData->get('fields') ?: [
				'transition',
				['parent', 'parent_id'],
				['published', 'state', 'enabled'],
				['category', 'catid'],
				'featured',
				'access',
				'id',
			];
`
old syntax:
 `			$fields = $displayData->get('fields') ?: array(
				'transition',
				array('parent', 'parent_id'),
				array('published', 'state', 'enabled'),
				array('category', 'catid'),
				'featured',
				'access',
				'id',
			);
`

hgh-esn replied:
Done

Quy commented on administrator/components/com_plugins/tmpl/plugins/default.php:
```diff
@@ -111,6 +111,11 @@
 									<?php echo $editIcon; ?><?php echo $item->name; ?></a>
+									<?php if (!empty($item->note) && trim($item->note) != '') : ?>
+										<div class="small">
+											<strong><?php echo JText::_('JFIELD_NOTE_LABEL') . ': '; ?></strong><?php echo $this->escape($item->note); ?>
```
```suggestion
											<strong><?php echo Text::_('JFIELD_NOTE_LABEL') . ': '; ?></strong><?php echo $this->escape($item->note); ?>
```

hgh-esn replied:
Done

hgh-esn committed:
changed language-key in alphabetical order

changed position of COM_PLUGINS_FIELD_NOTE_LABEL="Note" behind COM_PLUGINS_FIELD_NAME_LABEL="Plugin Name"

richard67 reviewed (changes requested):
Quy is right, it is not perfect yet.

hgh-esn reviewed (commented):
And why do I find often "JText_(xxxx" statements in Joomla and now it should be reduced to "Text_(" ?
Can anyone explain that?

hgh-esn committed:
change JText_( ... )   to Text_( ... )

hgh-esn committed:
changed "array( ... )" to short array syntax "[ ... ]"

brianteeman commented:
> And why do I find often "JText_(xxxx" statements in Joomla and now it should be reduced to "Text_(" ?
> Can anyone explain that?

Joomla 3 used JText
Joomla 4 uses Text

There is currently **one** instance of JText to be updated

hgh-esn commented:
@brianteeman 
Thanks for the explanation.

I think there were good reasons for that, but it makes the code not easier to read, because the assignment to Joomla is no longer directly recognizable.

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

richard67 commented:
@hgh-esn What happened? Clicked wrong button? If you did not close this PR intentionally, you should reopen it, because it is closed now.

hgh-esn commented:
@richard67 
you'r right, I clicked wrong button?

richard67 commented:
@mbabker Can you help with this PR not having a test item in the issue tracker?

hgh-esn commented:
@richard67 Can you please tell me what that is ... a test-item in the issue tracker?

richard67 commented:
@hgh-esn Issues and pull requests not only appear on GitHub but also on the Joomla issue tracker, [https://issues.joomla.org/](https://issues.joomla.org/).

For pull requests there is a button "Test this", which then allows someone who tested a PR to set the result accordingly.

It needs successful tests for a PR to get ready to be merged (ready to commit, RTC).

Sometimes this issue tracker has a problem with a PR, like it is with yours. The button for the test is missing in the issue tracker, so someone who tests your PR can not set the result there.

This is not your fault, it can happen sometimes with that issue tracker in recent times.

hgh-esn commented:
@richard67 thanks for that.

hgh-esn commented:
@richard67 

... PS: Can I contact you via your homepage(email) and later by phone.
I think it is better to talk about that here by phone.

richard67 commented:
@hgh-esn No, better not. If you need some more information about Joomla coding stuff, check documentation e.g. here: [https://docs.joomla.org/Git_for_Coders](https://docs.joomla.org/Git_for_Coders).

hgh-esn commented:
@mbabker 
Thanks for setting up he "Test this"  button.

@richard67 
You can use it now.

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

infograf768 commented on layouts/joomla/edit/global.php:
```diff
@@ -27,20 +27,35 @@
-	'note',
-	'version_note',
-);
+if ($component !== 'com_plugins')
```
Please explain why you modify this file.
I tested this PR without that modification and it looks like working fine.

infograf768 replied:
forget it. See below. :)

hgh-esn replied:
Yep, 
I think you were running into the same bug as I did  in early coding times.
I couldn't find out, why it was so with the double displayed  note-fields.
But I figured out the way with the code-switch ( if ... for others else ... for plugins).
So I was happy found out a solution.

richard67 replied:
but you should use positive logic, is a golden programming rule. I.e. check in the if for $component === 'com_plugins' and execute code for plugins there, and in the else do the other code. Got me?

richard67 commented:
@hgh-esn Regarding functionality and code style, your PR seems to be ok now. But there is an issue regarding consistency of UI compared to other note fields in the list views, from my point of view.

The note field of site modules is shown in the modules list as follows in a div, with text in brackets and no `<strong>` attribute:

![note_site-modules](https://user-images.githubusercontent.com/7413183/58694935-730ad900-8394-11e9-99f9-62cffdc962c4.png)

The note field of categories is shown in the categories list in the same way, but in a span instead of a div so it appears in the same line:

![note_categories](https://user-images.githubusercontent.com/7413183/58695093-d09f2580-8394-11e9-8a97-8d38776187bd.png)

For articles it is same as for categories except that there is a break-word class in addition:

![note_articles](https://user-images.githubusercontent.com/7413183/58695144-eb719a00-8394-11e9-8944-0999ee068c54.png)

So there is no 100% consistent way, but at least they have in common that nowhere is used `<strong>`.

With this PR, another new way to show the note in the list view is added for plugins:

![note_plugins](https://user-images.githubusercontent.com/7413183/58695212-165bee00-8395-11e9-83ef-60d53aaf61df.png)

I think it does not look bad, but it increases inconsistency in UI, so I suggest to change it so it is the same as with site modules.

But that's only my opinion.

@infograf768 What do you suggest?

brianteeman commented:
We must have a consistent UI

richard67 commented:
@brianteeman That was the intention of my comment, and that's why I wait with testing. But as you see we are not 100% consistent at other places, too. I suggest to do it in this PR in the same way as in the site modules, and clean up other inconsistencies with a separate PR. Do you agree? If not, what do you suggest?

infograf768 commented:
I rather like the way the Note is displayed here, i.e. below. It is much clearer to me.
We could, in another PR, normalize this placement.

Just found another bug.
Edit a Fields plugin. One gets 2 Note fields.

<img width="346" alt="Screen Shot 2019-05-31 at 11 16 15" src="https://user-images.githubusercontent.com/869724/58695618-20cab780-8396-11e9-8702-7091ebe9687c.png">

richard67 commented:
@infograf768 I think that was the reason for the change in file "layouts/joomla/edit/global.php". Did you test with this change included? I did, and I don't have that issue with duplicate note field in a fields plugin, e.g. "Fields - Integer".

richard67 commented:
@infograf768 And regarding IU: I agree with you that it looks better below than beside in the list view, but what about the `<strong>`, shouldn't that be removed? I think it would be ok if done like for the site modules.

infograf768 commented:
> I think that was the reason for the change in file "layouts/joomla/edit/global.php". Did you test with this change included? I did, and I don't have that issue with duplicate note field in a fields plugin, e.g. "Fields - Integer".

You are correct. 

> And regarding IU: I agree with you that it looks better below than beside in the list view, but what about the <strong>, shouldn't that be removed? I think it would be ok if done like for the site modules.

Agree too.

richard67 commented:
@infograf768 The fact that it is in brackets in the list view does not look nice. I think we should remove those brackets everywhere later in another PR.

Regarding the change in file "layouts/joomla/edit/global.php": I think it should not use negative logic, i.e. not check for `$component !== 'com_plugins'` but for `$component === 'com_plugins'` and exchange if and else code accordingly. Do you agree?

infograf768 commented:
Indeed, better positive than negative.
Not sure I understand about the brackets.

richard67 commented:
@infograf768 In site menus list: `(Note: blabla)`. In categories list: `(Alias: uncategorized, Note: blabla)`. I mean the `(` `)` brackets. See my screenshot above.

infograf768 commented:
Yep, once we modify the display of the Note to be below the title, the parentheses are unnecessary
😉  https://practicaltypography.com/parentheses-brackets-and-braces.html

richard67 commented:
Point is we should be consistent, and that means text JGLOBAL_LIST_NOTE should be used in the list view and not JFIELD_NOTE_LABEL in order to be consistent with the other list views. I will make a review with change request for that. Later we can change these texts for removing the brackets in another PR.

richard67 commented on administrator/components/com_plugins/tmpl/plugins/default.php:
```diff
@@ -111,6 +111,11 @@
 									<?php echo $editIcon; ?><?php echo $item->name; ?></a>
+									<?php if (!empty($item->note) && trim($item->note) != '') : ?>
+										<div class="small">
+											<strong><?php echo Text::_('JFIELD_NOTE_LABEL') . ': '; ?></strong><?php echo $this->escape($item->note); ?>
```
```suggestion
											<?php echo Text::sprintf('JGLOBAL_LIST_NOTE', $this->escape($item->note)); ?>
```

richard67 replied:
For consistent UI with other list views having notes, e.g. site modules.

richard67 replied:
See here for example: [https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_modules/tmpl/modules/default.php#L146](https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_modules/tmpl/modules/default.php#L146)

hgh-esn replied:
@richard67 @infograf768  
I disagree  with the reduction of the strong-tag and the ":". 
I was my opinion it is to better separate the "Note:" from it's value(text).
When the "Note:" isn't highlighted, it looks like being integrated into the value(text).
When this is not the way it is done in other ways, please lets harmonise the other notes to have a common look

richard67 replied:
@hgh-esn as Brian has mentioned, we have to be consistent in UI. I agree with you that your solution looks nice, but it is not consistent with other list views having note, as I have proven with my screen shots. So please accept that change and we can make the notes nice for all list vies with another, future pull request.

hgh-esn replied:
@richard67 

> <?php if (!empty($item->note) && trim($item->note) != '') : ?>
The `trim($item->note) != ''` I think is necessary, because it works as follows:

When a user enters a ' ' or many spaces without  other alpha-chars, it seams that the field has no content.  In the listview then there is a Note: .....  without viewable chars.
In the other notes it is as I described here. There it is - for my opinion - also necessary to change it.
With the `trim($item->note) != ''` there is no appearance of an empty note.

richard67 replied:
As I said, be consistent with other things having a note item, like the site modules. I posted a link where you could see the example.

richard67 replied:
@hgh-esn Ok, the `trim($item->note) != '' ` thing I don't mind, you can leave it in, but this change here you should do for consitent UI, even if your thing would look better. Otherwise I think Brian or JM (Jean-Marie = infograf768) may give you a negative test result. We can later makde display of all notes in list views better with a future PR. UI consistency is an important thing, you know.

hgh-esn replied:
Sorry, but when changing other note-viewing anyway, why not changing so as I worked it out?
Also !
You strain my nerves really with the basic questions here!

It was not my intention with this plugin-note field to start such a discussion here. I've been wondering for a long time why previous versions of these notes were so inconsistent and, moreover, just "forgotten" in the plugins.
As you can see, the "note theme" with its various solutions in Joomla so far is not all harmonic.

There is still much to do.

richard67 replied:
I strain your nerves? I would say you strained our nerves with the way how you worked, inlcuding all those file upload commits and so on. We provided code reviews and all here in order to help you to get your PR accepted. If you would have read documentation before, including code style guydelines, it would have been much easier for you and for us.

hgh-esn replied:
Please do not just present it as if the chaos only emanates from me. Yes, it was bumpy in the beginning - no question, but that is probably not spared anyone who goes on this terrain.
**I would like to express my gratitude to everyone who has helped to bring this about to an end.**
Incidentally, until yesterday afternoon I thought that everything was ok and that only the "test-this" button was missing. But what starts here this morning, that beats - from my point of view - the bottom of the barrel. In my opinion, all these fundamental questions could have been discussed and clarified much earlier. The code is here open and can be viewed at any time.
So it is incomprehensible to me, that here now the questions over
- pos./neg. logic
- Use of global language keys
- Output formats
- Harmonization with other notes in modules etc. (although, as stated, they are not really harmonious)
are being raised.

I will confirm the last requirements, but **hereby point out** that **I do not think they are beautiful**.
I hope I still get everything together.

richard67 commented on layouts/joomla/edit/global.php:
```diff
@@ -27,20 +27,35 @@
-	'note',
-	'version_note',
-);
+if ($component !== 'com_plugins')
```
Please change this if statement so it does not use negative logic, i.e. not check for `$component !== 'com_plugins'` but for `$component === 'com_plugins'` and exchange if and else code accordingly.

In addition, remove spaces at line end as suggested with the 2 changes below.

richard67 commented on layouts/joomla/edit/global.php:
```diff
@@ -27,20 +27,35 @@
+		'note',
+		'version_note',
+	];
+} 
```
```suggestion
}
```

richard67 commented on layouts/joomla/edit/global.php:
```diff
@@ -27,20 +27,35 @@
+		'version_note',
+	];
+} 
+else 
```
```suggestion
else
```

richard67 commented on administrator/components/com_plugins/tmpl/plugins/default.php:
```diff
@@ -111,6 +111,11 @@
 								<?php $editIcon = $item->checked_out ? '' : '<span class="fa fa-pen-square mr-2" aria-hidden="true"></span>'; ?>
 								<a class="hasTooltip" href="<?php echo Route::_('index.php?option=com_plugins&task=plugin.edit&extension_id=' . (int) $item->extension_id); ?>" title="<?php echo Text::_('JACTION_EDIT'); ?> <?php echo $this->escape(addslashes($item->name)); ?>">
 									<?php echo $editIcon; ?><?php echo $item->name; ?></a>
+									<?php if (!empty($item->note) && trim($item->note) != '') : ?>
```
```suggestion
									<?php if (!empty($item->note)) : ?>
```

richard67 replied:
If not empty should be enough, see as example the note field in the site modules list here:
[https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_modules/tmpl/modules/default.php#L144](https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_modules/tmpl/modules/default.php#L144)

hgh-esn replied:
@richard67 

> <?php if (!empty($item->note) && trim($item->note) != '') : ?>
The `trim($item->note) != ''` I think is necessary, because it works as follows:

When a user enters a ' ' or many spaces without  other alpha-chars, it seams that the field has no content.  In the listview then there is a Note: .....  without viewable chars.
In the other notes it is as I described here. There it is - for my opinion - also necessary to change it.
With the `trim($item->note) != ''` there is no appearance of an empty note in the listview.

richard67 replied:
As I said, be consistent with other things having a note item, like the site modules. I posted a link where you could see the example: [https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_modules/tmpl/modules/default.php#L146](https://github.com/joomla/joomla-cms/blob/4.0-dev/administrator/components/com_modules/tmpl/modules/default.php#L146)

richard67 commented:
Completed testing instructions<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24967">issues.joomla.org/tracker/joomla-cms/24967</a>.</sub>

hgh-esn committed:
Update layouts/joomla/edit/global.php

Co-Authored-By: Richard Fath <richard67@users.noreply.github.com>

hgh-esn committed:
Update layouts/joomla/edit/global.php

Co-Authored-By: Richard Fath <richard67@users.noreply.github.com>

richard67 commented on layouts/joomla/edit/global.php:
```diff
@@ -27,20 +27,35 @@
-	'note',
-	'version_note',
-);
+if ($component !== 'com_plugins')
```
Pease use positive logic, is a golden rule in programming:
```suggestion:-0+28
if ($component === 'com_plugins')
{
	$fields = $displayData->get('fields') ?: [
		'transition',
		['parent', 'parent_id'],
		['published', 'state', 'enabled'],
		['category', 'catid'],
		'featured',
		'access',
		'id',
	];
}
else
{
	$fields = $displayData->get('fields') ?: [
		'transition',
		['parent', 'parent_id'],
		['published', 'state', 'enabled'],
		['category', 'catid'],
		'featured',
		'sticky',
		'access',
		'id',
		'language',
		'tags',
		'note',
		'version_note',
	];
};
```

hgh-esn replied:
@richard67 
please don't hurry, I working me thru this explosion of comments step by step.

I know it and my opinion was to change the existing code in minimal ways,
But I fell free to change it.

hgh-esn replied:
Done

richard67 replied:
@hgh-esn It seems you made a mistake with that. Now following code lines are duplicate:

> if ($component === 'com_categories')
> {
> 	$extension = $input->getCmd('extension', 'com_content');
> 	$parts     = explode('.', $extension);
> 	$component = $parts[0];
> }
> 
> $saveHistory = ComponentHelper::getParams($component)->get('save_history', 0);
>

hgh-esn committed:
changed up to positive logic in if statement

if ($component === 'com_plugins')
.....
else
.....

richard67 commented on layouts/joomla/edit/global.php:
```diff
@@ -27,20 +27,44 @@
-	'note',
-	'version_note',
-);
+if ($component === 'com_categories')
```
Remove lines 30 to 37, they somehow became duplicated with your last commit.

hgh-esn replied:
Done, sorry for that.

hgh-esn committed:
cut off duplicate code from further editation

if ($component === 'com_categories')
{
....
}

hgh-esn committed:
Update administrator/components/com_plugins/tmpl/plugins/default.php

I don't really agree with that, but to go further on, I'll accept it.

Co-Authored-By: Richard Fath <richard67@users.noreply.github.com>

hgh-esn committed:
Update administrator/components/com_plugins/tmpl/plugins/default.php

I don't really agree with that, but to go further on, I'll accept it.

Co-Authored-By: Richard Fath <richard67@users.noreply.github.com>

infograf768 commented on administrator/components/com_plugins/forms/plugin.xml:
```diff
@@ -58,5 +58,15 @@
+		<field
+			name="note"
+			type="textarea"
+			label="COM_PLUGINS_FIELD_NOTE_LABEL"
```
Please use the Global String:
`JFIELD_NOTE_LABEL="Note"`
and delete `COM_PLUGINS_FIELD_NOTE_LABEL`

Note to @SharkyKZ Concerning the PR to normalise the Note display, we may have to modify at  the same time other similar instances in core as a Global string is always better in such cases. When doing this, we will have to delete unused specific strings such as `COM_CATEGORIES_FIELD_NOTE_LABEL` except maybe for Workflow.

@hgh-esn
Please understand that we are in J4 and some stuff is discovered when some new code is proposed. J4 is a good occasion to clean up stuff.
Thanks for your patience.

hgh-esn replied:
Done

infograf768 replied:
It is not done. You have not deleted COM_PLUGINS_FIELD_NOTE_LABEL

SharkyKZ commented:
Please revert changes to the `global.php` layout file. There should be no component-specific code like this. Instead, you can either remove the field from `edit.php` and have it rendered by `global.php`. Or set `$this->fields` property with form field names excluding `note`.

hgh-esn commented:
@SharkyKZ @infograf768 

Do not really know what  "**SharkyKZ**" means in details.
**I'm** confused and **out here** until you cleared it internally.

infograf768 commented:
@SharkyKZ means undo the change to global.php and modify edit.php to get something like
```
			<div class="col-md-3">
				<div class="card card-light">
					<div class="card-body">
						<?php
						// Set main fields.
						$this->fields = array(
							'enabled',
							'access',
						);

						echo LayoutHelper::render('joomla.edit.global', $this); ?>

						<div class="form-vertical form-no-margin">
							<div class="control-group">
								<div class="control-label">
									<?php echo $this->form->getLabel('ordering'); ?>
								</div>
								<div class="controls">
									<?php echo $this->form->getInput('ordering'); ?>
								</div>
							</div>
							<div class="control-group">
								<div class="control-label">
									<?php echo $this->form->getLabel('folder'); ?>
								</div>
								<div class="controls">
									<?php echo $this->form->getInput('folder'); ?>
								</div>
							</div>
							<div class="control-group">
								<div class="control-label">
									<?php echo $this->form->getLabel('element'); ?>
								</div>
								<div class="controls">
									<?php echo $this->form->getInput('element'); ?>
								</div>
							</div>
							<div class="control-group">
								<div class="control-label">
									<?php echo $this->form->getLabel('note'); ?>
								</div>
								<div class="controls">
									<?php echo $this->form->getInput('note'); ?>
								</div>
							</div>
						</div>
etc.
```

hgh-esn commented:
@infograf768  @SharkyKZ  @richard67 

That is exactly that what as it was in my early codings.
The note filed is here in `edit.php` as you can see at the end.
When reversing the `global.php` we go back to that, what you have seen this morning in your test.
The note input-field will then be double displayed. (I don't know why ?)
But with the changes in `global.php` the doubling effect has gone away.

@SharkyKZ
Please make me a good functional code-tip to get your idea to the right.

richard67 commented:
Is not related to your PR or the changes, is a general problem with AppVeyor it seems.

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

hgh-esn committed:
label changed to JFIELD_NOTE_LABEL

infograf768 commented:
@hgh-esn
test my code above after taking off the changes in global.php
I tested it here and the Note field is displayed only once, including for the Fields plugin.
```
<?php
						// Set main fields.
						$this->fields = array(
							'enabled',
							'access',
						);

						echo LayoutHelper::render('joomla.edit.global', $this); ?>
```
is taking care of that.

infograf768 commented on administrator/language/en-GB/en-GB.com_plugins.ini:
```diff
@@ -10,6 +10,7 @@ COM_PLUGINS_ERROR_FILE_NOT_FOUND="The file %s could not be found."
 COM_PLUGINS_FIELD_ELEMENT_LABEL="Plugin File"
 COM_PLUGINS_FIELD_FOLDER_LABEL="Plugin Type"
 COM_PLUGINS_FIELD_NAME_LABEL="Plugin Name"
+COM_PLUGINS_FIELD_NOTE_LABEL="Note"
```
DELETE as it is now useless.

hgh-esn commented:
> ```
> 						<?php
> 						// Set main fields.
> 						$this->fields = array(
> 							'enabled',
> 							'access',
> 						);
> 
> 						echo LayoutHelper::render('joomla.edit.global', $this); ?>
> ```
@infograf768 
Thanks for testing and your experiments with the code.

I'll do that changes at the afternoon. Now I have no time.
But one question there is.
Days ago, we had a discussion about using "php short array syntax".
When I rebuild the `global.php` should I use the standard php array syntax or the short array variant?
The same is in `edit.php`, for the new statement  `$this->fields = array('enabled','access',);?

I ask this, to avoid any an upcoming discussion about the, ;)

infograf768 commented:
When you “rebuild” global.php, just use the existing 4.0 global.php. it is a revert, not really a rebuild.

hgh-esn committed:
Set main fields.

hgh-esn committed:
delete key COM_PLUGINS_FIELD_NOTE_LABEL

hgh-esn committed:
revert to global.php in jooma-cms/4.0-dev

hgh-esn committed:
delete wrong code

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

hgh-esn commented:
@infograf768 @richard67  
Changes are done - testing can go further on.

infograf768 commented:
I have tested this item :white_check_mark: successfully on 2b77e9612e98393d3c9021088cdb5886baec3ad5<br /><br />OK now. I propose some small glitches to be modified later in another PR.<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24967">issues.joomla.org/tracker/joomla-cms/24967</a>.</sub>

hgh-esn commented:
@infograf768 
Thanks for that good news. 
It was a strongly birth to get it to run but, now - I hope - it seams ok.
Who will be the second tester? 
Enjoy a good sunny sunday.
<hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24967">issues.joomla.org/tracker/joomla-cms/24967</a>.</sub>

franz-wohlkoenig commented:
>  Who will be the second tester?

a Volunteer which take Time and is interestet to test.

hgh-esn commented:
@franz-wohlkoenig 

Does that mean, that if no one feels called, then the PR is hanging around here endlessly.
Can the PR creator also do a second test to push the PR and can he then be the second tester?

Note: In my previous company, there was a test team that carries out such activities. The creator was excluded from it.

franz-wohlkoenig commented:
> Does that mean, that if no one feels called, then the PR is hanging around here endlessly.

Correct. All are Volunteers.

>  Can the PR creator also do a second test to push the PR and can he then be the second tester?

Creator test can't count as the Project expect that the Pull Request is tested by Creator before creating the PR.

richard67 commented:
I have tested this item :white_check_mark: successfully on 2b77e9612e98393d3c9021088cdb5886baec3ad5<br /><br />Note field for plugins works.

Display in the list view is consistent with other list views having notes, e.g. site menus.

The display of notes values in list views is not nice in general and should be improved consistently for all list views with another, future PR. <hr /><sub>This comment was created with the <a href="https://github.com/joomla/jissues">J!Tracker Application</a> at <a href="https://issues.joomla.org/tracker/joomla-cms/24967">issues.joomla.org/tracker/joomla-cms/24967</a>.</sub>

richard67 commented:
@hgh-esn The reason why you had to run through many code reviews was mainly caused by your implementation and especially at the beginning in the way how you worked, that you committed code which obviously could not work, like e.g. the postgresql stuff.

As Franz mentioned above, we expect that when someone proposes a PR, he has already tested his PR so it will not contain failures.

Beside this, we (volunteers) have to make sure that Joomla code has a certain quality, and so there are code style rules and architectural aspects to be mentioned. It is not enough if a new code just functionally does what it shall do. The code has to have the required quality, UI consistency is also an aspect.

Not everbody of us has the same skills and focus, and so it can happen that one of us (e.g. me) is happy with your PR, but then someone who knows more about Joomla architecture than I do (e.g. SharkyKZ) finds another thing to be improved.

And so yes, it can be hard at the beginning to go through all those reviews, but it is required because we want to have well tested code in a good quality so it is better maintainable.

After a while, people get familiar with that and it will be easier.

Regarding code style you can find some documentation here:

[https://developer.joomla.org/coding-standards/basic-guidelines.html](https://developer.joomla.org/coding-standards/basic-guidelines.html)

But it also possible to check the log files of drone when after a commit, drone failed. If section phpcs is red, there are code style errors, and a click on that section will show the log file.

In general some links to further documents about contributing to Joomla can also be found on this German start page:

[https://docs.joomla.org/Portal:Joomla!_Code_Contributors/de](https://docs.joomla.org/Portal:Joomla!_Code_Contributors/de).

franz-wohlkoenig commented:
Status "Ready To Commit".

infograf768 committed:
Merge branch '4.0-dev' into 4.0-dev

hgh-esn commented:
@richard67 

In your previous explanation, of course, you are right, but allow me some personal views on this:

1. Central Joomla Code:  'global.php'
It would be very nice to read more about, how it is used properly . Since I did not see the exact function there, I got to the solution with the "code-switch" (if .... === com_plugin)
then .... else ....;
resorted.

That this was also different, as SharkeyKZ has described, I could have found synonymous, if I had checked in module notes. Unfortunately I have not - my mistake.

But as I said, all that could be avoided, if one would document a bit more in the header of `global.php`, what is done there by default and how one can possibly influence the behavior from the outside.
Such inline comments can often be very helpful.

2. .... we expect that when someone proposes a PR, he has already tested his PR so it will not contain failures.

I tested everything in my personal environment before handing it over to you. It worked for me, there was - except for this small DB definition problem - no really functional errors. The note could be displayed in both affected views.

Ok, everything that has developed after that is not that great and probably had to come like that. Some discussions (short-array-code, etc.) were - in my opinion - but also unnecessary. This spelling may be for expert programmers, but it does not make the code more readable.
If you load such a code into an editor with syntax highlighting, you will not be happier.

But "sponge over it", that's all now past.

If you want to tackle the topic: notes and their mapping in Joomla again, I offer my further assistance.

I think we should first
- find out where the affected areas are.
- Then you have to specify what the appearance of the note (s) in Joomla should look like.
- only then should the technical implementation take place.

Should I open an issue in J4?

richard67 commented:
Why closed? It was RTC, means ready to commit. You hit the wrong button accidently?

richard67 commented:
@hgh-esn Why closed? It was RTC, means ready to commit. You hit the wrong button accidently?

hgh-esn commented:
Oh sorry for that !

richard67 commented:
@franz-wohlkoenig Could you add back RTC label? Author has accidently closed the PR.

franz-wohlkoenig commented:
Status "Ready To Commit".

hgh-esn commented:
@infograf768  @richard67 

I try to setup a joomla environment to test the PR myself.
I worked as described in this [dok](https://docs.joomla.org/Testing_Joomla!_patches/en#).
I made the download from the joomla staging-environment and have that installed. It runs, but it is a **J3.9.7-dev** environment.
And here is my question:
Do I have to migrate from **J3.9.7-dev** to **J4-dev** first?
If so, that isn't described in that doku, or I haven't found anything about it. ;)

What is your install-process for that patch in details?

infograf768 committed:
Merge branch '4.0-dev' into 4.0-dev

infograf768 commented:
![4 0](https://user-images.githubusercontent.com/869724/58777891-b9e51280-85d0-11e9-898e-0a4c9628a4e4.png)

hgh-esn commented:
@infograf768 

I did that also. But I get this when starting the backend.
<img width="908" alt="2019-06-03_07h32_18" src="https://user-images.githubusercontent.com/5299726/58778219-ffeea600-85d1-11e9-9b13-ec6c7a3999c0.png">

infograf768 commented:
That is normal. One has to use Composer and NPM. This PR is not the place to explain the HOWTO.

infograf768 committed:
Merge branch '4.0-dev' into 4.0-dev

richard67 commented:
@hgh-esn Find here a link to a document in German language about how to set up your development environment for J4, including the composer and npm stuff which Jean-Marie mentioned:
[https://docs.joomla.org/J4.x:Setting_Up_Your_Local_Environment/de](https://docs.joomla.org/J4.x:Setting_Up_Your_Local_Environment/de)

Note also that you should not try to update current staging or 3.9.x to 4.0 with the Joomla Update function. This currently is not intended to work. Later update from 3.10 to 4.0 will me made possible.

infograf768 committed:
Merge branch '4.0-dev' into 4.0-dev

hgh-esn committed:
Merge branch '4.0-dev' into 4.0-dev

hgh-esn commented:
@richard67  Thanks for the link, but that is quit different to that, what is written down in  this [dok](https://docs.joomla.org/Testing_Joomla!_patches/en#). And all together it looks not as easy as a normal tester is willing it to. 
It looks very abstract, because of having to install a lot of additional software, with their handling / parameterization a normal tester may be a bit overwhelmed.
At any rate, it was not even fast. I'll just wait for the "Nightly Build Version", where I can see the result. 

Allow me one more question:
When will the PR be transferred to the J4 code? "Ready to comit" is it now.

Does anyone have to do anything or is this automatic?

franz-wohlkoenig commented:
@hgh-esn thiswill be done by Release Lead @wilsonge if he decide to merge the PR. A RTC is no automatic that a PR get in core.

infograf768 committed:
Merge branch '4.0-dev' into 4.0-dev

infograf768 commented:
@wilsonge
Can you take a decision on this one?

wilsonge commented:
Thanks! Sorry for the delay merging
